minorminer/utils/zephyr/node_edge.py

```diff
@@ -0,0 +1,796 @@
+        self._edge = self._set_edge(x, y)
+
+    def _set_edge(self, x, y):
+        """Returns ordered tuple corresponding to the set {x, y}."""
+        if x < y:
+            return (x, y)
+        else:
+            return (y, x)
+
+    def __hash__(self):
+        return hash(self._edge)
+
+    def __getitem__(self, index: int) -> int:
+        return self._edge[index]
+
+    def __eq__(self, other: Edge):
+        return self._edge == other._edge
+
+    def __str__(self) -> str:
+        return f"{type(self).__name__}{self._edge}"
+
+    def __repr__(self) -> str:
+        return f"{type(self).__name__}{self._edge}"
+
+
+class ZEdge(Edge):
+    """Initializes a ZEdge with 'ZNode' nodes x, y.
+
+    Args:
+        x (ZNode): Endpoint of edge. Must have same shape as ``y``.
+        y (ZNode): Endpoint of edge. Must have same shape as ``x``.
+        check_edge_valid (bool, optional): Flag to whether check the validity of values and types of ``x``, ``y``.
+            Defaults to True.
+
+    Raises:
+        TypeError: If either of x or y is not 'ZNode'.
+        ValueError: If x, y do not have the same shape.
+        ValueError: If x, y are not neighbors in a perfect yield (quotient)
+            Zephyr graph.
+
+    Example 1:
+    >>> from zephyr_utils.node_edge import ZNode, ZEdge
+    >>> e = ZEdge(ZNode((3, 2)), ZNode((7, 2)))
+    >>> print(e)
+    ZEdge(ZNode(CartesianCoord(x=3, y=2, k=None)), ZNode(CartesianCoord(x=7, y=2, k=None)))
+    Example 2:
+    >>> from zephyr_utils.node_edge import ZNode, ZEdge
+    >>> ZEdge(ZNode((2, 3)), ZNode((6, 3))) # raises error, since the two are not neighbors
+    """
+
+    def __init__(
+        self,
+        x: ZNode,
+        y: ZNode,
+        check_edge_valid: bool = True,
+    ) -> None:
+        if check_edge_valid:
+            if not isinstance(x, ZNode) or not isinstance(y, ZNode):
+                raise TypeError(f"Expected x, y to be ZNode, got {type(x), type(y)}")
+
+            if x.shape != y.shape:
+                raise ValueError(f"Expected x, y to have the same shape, got {x.shape, y.shape}")
+
+            for kind in EdgeKind:
+                if x.is_neighbor(y, nbr_kind=kind):
+                    self._edge_kind = kind
+                    break
+            else:
+                raise ValueError(f"Expected x, y to be neighbors, got {x, y}")
+
+        self._edge = self._set_edge(x, y)
+
+    @property
+    def edge_kind(self) -> EdgeKind:
+        return self._edge_kind
+
+
+class ZNode:
+    """Initializes 'ZNode' with coord and optional shape.
+
+    Args:
+        coord (CartesianCoord | ZephyrCoord | tuple[int]): Coordinate in (quotient) Zephyr or (quotient) Cartesian
+        shape (ZShape | tuple[int | None] | None, optional): Shape of the Zephyr graph containing this ZNode.
+            If a ZShape is passed, it should be a namedtuple with fields `m` (grid size of the Zephyr graph) and `t` (tile size of the Zephyr graph).
+        Defaults to None.
```

Comment:
```suggestion
        coord (CartesianCoord | ZephyrCoord | tuple[int]): Coordinate in (quotient) Zephyr or (quotient) Cartesian.
        shape (ZShape | tuple[int | None] | None, optional): Shape of the Zephyr graph containing this ZNode.
            If a ZShape is passed, it should be a namedtuple with fields `m` (grid size of the Zephyr graph) 
            and `t` (tile size of the Zephyr graph). Defaults to None.
```
Just a bit of docstring reformatting.